src/%IPM/Main.cls

```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+		If (resultSet.%Get("Type") = "F") {
+			// get the exact version string from filename
+			Set ipmLocalArray($Piece($Piece(fileName, "ipm-", 2), ".xml")) = fileName
+			Set counter = counter + 1
+			Set menuList(counter) = $Piece($Piece(fileName, "ipm-", 2), ".xml")
+		}
+	}
+	$$$ThrowOnError(sc)
+
+	If (localOnly && (counter = 0)) {
+		$$$ThrowOnError($$$ERROR($$$GeneralError,"No ipm-<version>.xml installer file is found locally."))
+	}
+
+	If (version '= "") {
+		Set targetVersion = version
+		If '$DATA(ipmLocalArray(version)) {
+			If localOnly {
+				$$$ThrowOnError($$$ERROR($$$GeneralError,"No ipm-"_version_".xml installer file is found in "_XMLDir_" locally."))
+			} ElseIf 'quiet {
+				Write !, "No ipm-"_version_".xml installer file is found in "_XMLDir_" locally. Trying to download from the registry."
+			}
+			// need to get from remote registry server
+			Set useLocal = 0
+		}
+	} Else {
+		If quiet {
+			$$$ThrowOnError($$$ERROR($$$GeneralError,"Cannot install IPM in non-prompt mode without a specified IPM version."))
+		}
+		If (counter = 0) {
+			$$$ThrowOnError($$$ERROR($$$GeneralError,"No IPM version is found in "_XMLDir_" locally. Use version modifier to specify a version to download from the registry."))
+		}
+		// prompt user with local options
+		Set menuValue = 1
+ 		Set menuResponse = ##class(%Library.Prompt).GetMenu("Which version of local IPM to use?", .menuValue, .menuList,,128)
+		If (menuResponse '= 1) {
+			$$$ThrowOnError($$$ERROR($$$GeneralError,"Unsupported response. Operation cancelled."))
+		}
+		Set targetVersion = menuList(menuValue)
+	}
+	// Now we know which exact version to install and we know whether we should install it with local artifact or get from registry
+	
+	// Now, figure out which namespaces to install
+	If (quiet && (namespaces = "") && 'globally) {
+		$$$ThrowOnError($$$ERROR($$$GeneralError,"Cannot install IPM in non-prompt mode without specifying namespaces to install."))
+	}
+	If ((namespaces '= "") && globally) {
+		$$$ThrowOnError($$$ERROR($$$GeneralError,"Cannot specify namespaces and global installation flag at the same time."))
+	}
+
+	// Get all namespaces that currently do not have IPM enabled
+
+	// TODO: Let user interactively choose which namespaces to install IPM
+	// TODO: Better and more stable way to check whether IPM has been enabled for a namespace instead of checking IPM classes
+	Set baseNS = $Namespace
+	New $Namespace
+	Set $Namespace = "%SYS"
+	Set sc = ##class(%SYS.Namespace).ListAll(.allNS)
+	$$$ThrowOnError(sc)
+	Kill namespacesWithoutIPM		// multi-dim array to store all non-%SYS explicit namepaces that currently do not have IPM installed
+	Set currentNS=$ORDER(allNS(""))
+	While (currentNS '="") {
+		// avoid %SYS and implicit namespaces
+		If (currentNS '= "%SYS") && ($ZCONVERT(currentNS, "U") = currentNS) {
+			Set $Namespace = currentNS
+			If '$System.CLS.IsMthd("%IPM.Main", "Shell") {
+				Set namespacesWithoutIPM(currentNS) = 1
```

Comment:
Should also accumulate an array of namespaces with IPM and the version of IPM in those namespaces.
Based on namespaces selected/globally, may need to also upgrade IPM in those namespaces to match the provided version